src/hotspot/share/opto/loopopts.cpp

```diff
@@ -583,7 +583,9 @@ Node* PhaseIdealLoop::remix_address_expressions(Node* n) {
             n23_loop == n_loop) {
           Node* add1 = new AddPNode(n->in(1), n->in(2)->in(2), n->in(3));
           // Stuff new AddP in the loop preheader
-          register_new_node(add1, n_loop->_head->as_Loop()->skip_strip_mined(1)->in(LoopNode::EntryControl));
+          Node* entry = n_loop->_head->is_Loop() ? n_loop->_head->as_Loop()->skip_strip_mined(1)->in(LoopNode::EntryControl)
```

Comment:
Should we add a sanity assert (which could also serve as an explaining comment) that in the `RegionNode` case, it must be an irreducible loop head?

Reply:
I could do that but it would feel a bit arbitrary to put the assert only in this code, given that we have the same `is_Loop` check in a lot of other code as well. Maybe it would make sense to factor the checks into a separate method at some point.

Reply:
I agree with refactoring in separate RFE. 

Regarding these changes: should we simply skip this optimization if we see irreducible loop? They may have additional entry into the body so such transformation could be incorrect.

Reply:
Right, let's better be safe than sorry and skip this optimization for irreducible loops. The other optimizations in that method look safe though.